src/main/java/org/zalando/nakadi/service/kpi/NakadiPublisher.java

```diff
@@ -0,0 +1,127 @@
+        this.eventPublisher = eventPublisher;
+        this.uuidGenerator = uuidGenerator;
+        this.featureToggleService = featureToggleService;
+        this.batchCollectionTimeout = batchCollectionTimeout;
+        this.batchSize = batchSize;
+        this.pollTimeout = pollTimeout;
+        this.eventsQueueSize = eventsQueueSize;
+        this.scheduleTimeout = scheduleTimeout;
+        this.eventTypeEvents = new ConcurrentHashMap<>();
+        this.executorService = Executors.newFixedThreadPool(workers);
+        this.executorService.submit(() -> scheduleEventsBatchesSend());
+    }
+
+    private void scheduleEventsBatchesSend() {
+        LOG.trace("Schedule events collection on {}", Thread.currentThread().getName());
+        if (featureToggleService.isFeatureEnabled(FeatureToggleService.Feature.KPI_COLLECTION)) {
+            for (final String etName : eventTypeEvents.keySet()) {
+                executorService.submit(() -> sendEventBatch(etName));
+            }
+        }
+        try {
+            Thread.sleep(scheduleTimeout);
+        } catch (final InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
+        executorService.submit(() -> scheduleEventsBatchesSend());
+    }
+
+    private void sendEventBatch(final String etName) {
+        LOG.trace("Collecting batch for {} on {}", etName, Thread.currentThread().getName());
+        try {
+            final BlockingQueue<JSONObject> events = eventTypeEvents.get(etName);
+            final long finishAt = System.currentTimeMillis() + batchCollectionTimeout;
+            int eventsCount = 0;
+            JSONArray jsonArray = null;
+            while (eventsCount != batchSize && System.currentTimeMillis() < finishAt) {
+                try {
+                    final JSONObject event = events.poll(pollTimeout, TimeUnit.MILLISECONDS);
+                    if (event != null) {
+                        if (jsonArray == null) {
+                            jsonArray = new JSONArray();
+                        }
+                        jsonArray.put(event);
+                        eventsCount++;
+                    }
+                } catch (final InterruptedException ie) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+            try {
+                if (eventsCount == 0) {
+                    LOG.trace("No kpi events send to {}", etName);
+                    return;
+                }
+                eventPublisher.publish(jsonArray.toString(), etName, null);
+                LOG.trace("Published batch of {} to {}", eventsCount, etName);
+            } catch (final Exception e) {
+                LOG.error("Error occurred while publishing events to {}, {}", etName, e.getMessage(), e);
+            }
+        } finally {
+            if (featureToggleService.isFeatureEnabled(FeatureToggleService.Feature.KPI_COLLECTION)) {
+                executorService.submit(() -> sendEventBatch(etName));
```

Comment:
I don't exactly get how it works.
It submits itself for execution but at the same time it is also submitted for execution from `scheduleEventsBatchesSend` method.